packages/round-manager/src/context/RoundContext.tsx

```diff
@@ -0,0 +1,138 @@
+}
+
+interface Action {
+  type: ActionType;
+  payload?: any;
+}
+
+type Dispatch = (action: Action) => void;
+
+export const initialRoundState: RoundState = {
+  data: [],
+  isLoading: false,
+};
+
+export const RoundContext = createContext<
+  { state: RoundState; dispatch: Dispatch } | undefined
+>(undefined);
+
+const fetchRounds = async (
+  dispatch: Dispatch,
+  address: string,
+  walletProvider: Web3Provider,
+  programId: string
+) => {
+  dispatch({ type: ActionType.START_LOADING });
+
+  const res = await listRounds(address, walletProvider, programId);
+
+  if (res.error) {
+    dispatch({ type: ActionType.SET_LIST_ROUNDS_ERROR, payload: res.error });
+  }
+  dispatch({ type: ActionType.SET_ROUNDS, payload: res.data });
+  dispatch({ type: ActionType.FINISH_LOADING });
+};
+
+const fetchRoundById = (
+  dispatch: Dispatch,
+  walletProvider: Web3Provider,
+  roundId: string
+) => {
+  dispatch({ type: ActionType.START_LOADING });
+
+  getRoundById(walletProvider, roundId)
+    .then((round) =>
+      dispatch({ type: ActionType.SET_ROUNDS, payload: [round] })
+    )
+    .catch((error) =>
+      dispatch({ type: ActionType.SET_LIST_ROUNDS_ERROR, payload: error })
+    )
+    .finally(() => dispatch({ type: ActionType.FINISH_LOADING }));
+};
+
+const roundReducer = (state: RoundState, action: Action) => {
+  switch (action.type) {
+    case ActionType.START_LOADING:
+      return { ...state, isLoading: true };
+    case ActionType.FINISH_LOADING:
+      return { ...state, isLoading: false };
+    case ActionType.SET_ROUNDS:
+      return { ...state, data: action.payload ?? [] };
+    case ActionType.SET_LIST_ROUNDS_ERROR:
+      return { ...state, data: [], error: action.payload };
+  }
+};
+
+export const RoundProvider = ({ children }: { children: React.ReactNode }) => {
+  const [state, dispatch] = useReducer(roundReducer, initialRoundState);
+
+  const providerProps = {
+    state,
+    dispatch,
+  };
+
+  return (
+    <RoundContext.Provider value={providerProps}>
+      {children}
+    </RoundContext.Provider>
+  );
+};
+
+export const useRounds = (programId?: string) => {
+  const context = useContext(RoundContext);
+  if (context === undefined) {
+    throw new Error("useRounds must be used within a RoundProvider");
+  }
+  const { address, provider } = useWallet();
+
+  useEffect(() => {
+    if (programId) {
+      fetchRounds(context.dispatch, address, provider, programId);
+    }
+  }, [address, provider, programId, context.dispatch]);
+
+  return { ...context.state, dispatch: context.dispatch };
+};
+
+export const useRoundById = (roundId?: string) => {
+  const context = useContext(RoundContext);
+  if (context === undefined) {
+    throw new Error("useRounds must be used within a RoundProvider");
+  }
+  const { address, provider } = useWallet();
+
+  useEffect(() => {
+    if (roundId) {
+      const existingRound = context.state.data.find(
+        (round) => round.id === id
```

Comment:
my bad, used the wrong variable name in the previous suggestion 🙇‍♀️ 
```suggestion
        (round) => round.id === roundId
```